ci/snapcraft.sh

```diff
@@ -0,0 +1,135 @@
+# SNAP_EDGE=1 <-- with this, we build an edge/devel snap, otherwise a release snap
+# SNAP_ARCH=... <-- without this we cannot built, this is the snap architecture we're building
+
+if ! [ -n "$DO_SNAP" ]; then
+  echo "SNAP: Not attempting to snap"
+  exit 0
+fi
+
+if ! hash snapcraft >/dev/null 2>/dev/null; then
+  echo "SNAP: Unable to find snapcraft, cannot run"
+  exit 1
+fi
+
+if ! [ -n "$SNAP_ARCH" ]; then
+  echo "SNAP: Unable to generate snap, architecture unset"
+  exit 1
+else
+  echo "SNAP: Constructing snap for architecture $SNAP_ARCH"
+fi
+
+if [ -n "$SNAP_EDGE" ]; then
+  echo "SNAP: Generated snap will be an grade:devel snap"
+  GRADE=devel
+else
+  echo "SNAP: Generated snap will be a grade:stable snap"
+  GRADE=stable
+fi
+
+
+# This is the list of proxies which the snap will contain.  This needs to match
+# the set of aliases which the snap has been allocated.  If we add to this
+# list but don't get the aliases added, then people won't see the proxy except
+# as rustup.$PROXY.  If we have an alias which is not a supported proxy name
+# then rustup might get sad.
+
+PROXIES="cargo cargo-clippy cargo-fmt cargo-miri clippy-driver rls rustc rustdoc rustfmt"
+
+# From now on, things should be automagic
+
+VERSION=$(grep "^version" Cargo.toml | head -1 | cut -d\" -f2)
+
+rm -rf snap
+
+mkdir -p snap/snap
+
+cd snap
+
+cat > snap/snapcraft.yaml <<SNAP
+name: rustup
+version: $VERSION
+confinement: classic
+base: core18
+architectures:
+  - build-on: [amd64]
+    run-on: [$SNAP_ARCH]
+description: |
+  Rustup is the Rust Language's installer and primary front-end.  You probably
+  want this if you want to develop anything written in Rust.
+SNAP
+
+if [ -n "$SNAP_EDGE" ]; then
+  cat >> snap/snapcraft.yaml <<SNAP
+
+  Please note, this snap is experimental and functionality cannot be guaranteed
+  to be consistent with released snaps.
+summary: "EXPERIMENTAL: The Rust Language Installer"
+grade: devel
+SNAP
+else
+  cat >> snap/snapcraft.yaml <<SNAP
+summary: "The Rust Language Installer"
+grade: stable
+SNAP
+fi
+
+cat >> snap/snapcraft.yaml <<SNAP
+parts:
+  acquire:
+    plugin: dump
+    source: inputs
+    prime:
+     - "-rustup-init"
+  link:
+    plugin: nil
+    build-attributes: [no-patchelf]
+    after: [acquire]
+    override-pull: |
+      cp \$SNAPCRAFT_STAGE/rustup-init .
+    override-build: |
+      mkdir -p \$SNAPCRAFT_PART_INSTALL/bin
+      cp rustup-init \$SNAPCRAFT_PART_INSTALL/bin/rustup
+SNAP
+
+for PROXY in $PROXIES; do
+  cat >> snap/snapcraft.yaml <<SNAP
+      ln \$SNAPCRAFT_PART_INSTALL/bin/rustup \$SNAPCRAFT_PART_INSTALL/bin/$PROXY
+SNAP
+done
+
+cat >> snap/snapcraft.yaml << 'SNAP'
+    prime:
+     - "-rustup-init"
+
+environment:
+  RUSTUP_HOME: "$SNAP_USER_DATA/rustup"
+
+apps:
+  rustup:
+    command: bin/rustup
+SNAP
+
+for PROXY in $PROXIES; do
+  cat >> snap/snapcraft.yaml <<SNAP
+  $PROXY:
+    command: bin/$PROXY
+SNAP
+done
+
+mkdir inputs
+cp ../target/$TARGET/release/rustup-init inputs/
```

Comment:
Double quote `"$TARGET"` to avoid word splitting.

Reply:
Done.  Thanks